pennylane/utils.py

```diff
@@ -393,3 +393,54 @@ def expand_vector(vector, original_wires, expanded_wires):
     )
 
     return qml.math.reshape(expanded_tensor, 2**M)
+
+
+def prod_sort(op_list, wire_map: dict = None):
+    """Insertion sort algorithm that sorts a list of product factors by their wire indices, taking
+    into account the operator commutivity.
+
+    Args:
+        op_list (List[.Operator]): list of operators to be sorted
+        wire_map (dict): Dictionary containing the wire values as keys and its indexes as values.
+            Defaults to None.
+
+    Returns:
+        List[.Operator]: sorted list of operators
+    """
+
+    def swappable_ops(op1, op2, wire_map: dict) -> bool:
+        """Boolean expression that indicates if op1 and op2 are commutative and should be swapped when
+        sorting them by wire values.
+
+        Args:
+            op1 (.Operator): First operator.
+            op2 (.Operator): Second operator.
+            wire_map (dict): Dictionary containing the wire values as keys and its indexes as values.
+                Defaults to None.
+
+        Returns:
+            bool: True if operators should be swapped, False otherwise.
+        """
+        wires1 = op1.wires
+        wires2 = op2.wires
+        if wire_map is not None:
+            wires1 = wires1.map(wire_map)
+            wires2 = wires2.map(wire_map)
+        if np.intersect1d(wires1, wires2).size != 0:
+            return False
+        return np.min(wires1) > np.min(wires2)
+
+    if isinstance(op_list, tuple):
+        op_list = list(op_list)
+
+    for i in range(1, len(op_list)):
+
+        left_op = op_list[i]
+
+        j = i - 1
+        while j >= 0 and swappable_ops(op1=op_list[j], op2=left_op, wire_map=wire_map):
+            op_list[j + 1] = op_list[j]
+            j -= 1
+        op_list[j + 1] = left_op
```

Comment:
Just wondering: have you tried using the built-in `sorted()` function with `swappable_ops` as the key? It seems like there is a way to make it work and avoid all of this. (I think you would need to switch the True and False values however.)

Reply:
I just tried:
```
def swappable_ops(op1, op2, wire_map: dict=None) -> bool:
    wires1 = op1.wires
    wires2 = op2.wires
    if wire_map is not None:
        wires1 = wires1.map(wire_map)
        wires2 = wires2.map(wire_map)
    if np.intersect1d(wires1, wires2).size != 0:
        return 0
    return np.min(wires1) - np.min(wires2)

key = functools.cmp_to_key(swappable_ops)

sorted(ops, key=key)
```
And surprisingly, it's an order of magnitude slower... Usually in-built algorithms offer greater efficiency.

Reply:
Interesting 🤔 , I also thought it would be faster.

Reply:
`sorted()` uses a `divide and conquer` algorithm, which doesn't work in this case.

Counter-example:
```
op_list = [Y(2), CNOT(1, 2), Y(0), CNOT(0, 1)]

sorted_left = [Y(2), CNOT(1, 2)]
sorted_right = [Y(0), CNOT(0, 1)]
```
1st iteration (compare Y(0) and Y(2) ):
```
sorted_list = [Y(0)]

sorted_left = [Y(2), CNOT(1, 2)]
sorted_right = [CNOT(0, 1)]
```
2nd iteration (compare CNOT(0, 1) and Y(2)):
```
sorted_list = [Y(0), CNOT(0, 1)]

sorted_left = [Y(2), CNOT(1, 2)]
sorted_right = []
```
Result:
```
sorted_list = [Y(0), CNOT(0, 1), Y(2), CNOT(1, 2)]
```
But this is incorrect because CNOT(0, 1) and CNOT(1, 2) should not be swapped because they have overlapping wires.

Reply:
According to [the docs](https://docs.python.org/3/library/functions.html#sorted), the `sorted()` function is stable, so two elements will retain the original order if they compare equal (which is true for two operators with overlapping wires). I just tried @albi3ro's code on your `op_list` and it produces:
```python
[PauliY(wires=[0]), PauliY(wires=[2]), CNOT(wires=[1, 2]), CNOT(wires=[0, 1])]
```
which looks correct to me.

Reply:
Also @albi3ro, can you give an example that shows the slowdown? I timed yours and @AlbertMitjans's code on a few lists and got similar times (10000 loops ran):
```
time1 14.948776245117188
sorted1 [CNOT(wires=[0, 1]), PauliY(wires=[0]), PauliX(wires=[1]), PauliX(wires=[2]), PauliZ(wires=[0]) @ PauliY(wires=[2])]
time2 14.914478778839111
sorted2 [CNOT(wires=[0, 1]), PauliY(wires=[0]), PauliX(wires=[1]), PauliX(wires=[2]), PauliZ(wires=[0]) @ PauliY(wires=[2])]
```

Reply:
Weird... Running the following code:

```python
import functools

import pennylane as qml
from pennylane import numpy as np


def swappable_ops(op1, op2, wire_map: dict = None) -> bool:
    wires1 = op1.wires
    wires2 = op2.wires
    if wire_map is not None:
        wires1 = wires1.map(wire_map)
        wires2 = wires2.map(wire_map)
    if np.intersect1d(wires1, wires2).size != 0:
        return 0
    return np.min(wires1) - np.min(wires2)


ops = [qml.PauliY(2), qml.CNOT((1, 2)), qml.PauliY(0), qml.CNOT((0, 1))]
key = functools.cmp_to_key(swappable_ops)

print(sorted(ops, key=key))
```
I get:
```
[PauliY(wires=[0]), CNOT(wires=[0, 1]), PauliY(wires=[2]), CNOT(wires=[1, 2])]
```

Reply:
That's weird 🤔. I can't reproduce my previous output anymore.

Diving further into this, it looks like the `sorted()` version works for some permutations of the list and not others. I'm guessing it's because the `swappable_ops` key isn't transitive, e.g. `qml.PauliX(0)` is equal to `qml.PauliX(0) @ qml.PauliY(1)` which is equal to `qml.PauliY(1)`, but `qml.PauliX(0)` is not itself equal to `qml.PauliY(1)`. And the sorting behaviour isn't defined for a non-transitive binary relation.

So, it looks like the `sorted()` function is the wrong way to go here, and it's _probably_ fine to leave it as it is. Marking this as resolved.